File: apps/android/core/src/test/kotlin/com/intellieffect/drop/core/AuthStoreTest.kt
package com.intellieffect.drop.core

import java.time.Instant
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.test.runTest

/** iOS `AuthStoreTests`의 이식본. */
class AuthStoreTest {
    private val user = DropUser(id = "u1", email = "bruce@intellieffect.com")

    private fun session(user: DropUser = this.user) = Session(
        accessToken = "access",
        refreshToken = "refresh",
        expiresAt = Instant.ofEpochSecond(2_000_000_000),
        user = user,
    )

    private class FakeGateway(var restored: Session? = null) : AuthGateway {
        var signInResult: Result<Session> = Result.success(
            Session("a", "r", Instant.ofEpochSecond(2_000_000_000), DropUser("u1", null)),
        )
        var restoreError: Throwable? = null
        var signOutCount = 0
        var signOutError: Throwable? = null

        override suspend fun restore(): Session? {
            restoreError?.let { throw it }
            return restored
        }

        override suspend fun signIn(identity: GoogleIdentity): Session = signInResult.getOrThrow()

        override suspend fun signOut() {
            signOutCount += 1
            signOutError?.let { throw it }
        }
    }

    private class FakeIdentityProvider(
        var identity: GoogleIdentity? = GoogleIdentity(idToken = "id-token"),
    ) : GoogleIdentityProvider {
        var error: Throwable? = null
        var signOutCount = 0

        override suspend fun signIn(): GoogleIdentity? {
            error?.let { throw it }
            return identity
        }

        override fun signOut() {
            signOutCount += 1
        }
    }

    private fun store(
        gateway: FakeGateway = FakeGateway(),
        provider: FakeIdentityProvider = FakeIdentityProvider(),
    ) = Triple(AuthStore(gateway, provider), gateway, provider)

    /** 세션을 확인하기 전에는 로그인 화면도, 홈도 띄우지 않는다. */
    @Test
    fun `처음 상태는 미결정이다`() {
        val (store, _, _) = store()

        assertIs<AuthState.Undetermined>(store.state.value)
        assertNull(store.user)
    }

    @Test
    fun `저장된 세션이 있으면 로그인 상태로 복원된다`() = runTest {
        val (store, _, _) = store(FakeGateway(restored = session()))

        store.restore()

        assertEquals(user, (store.state.value as AuthState.SignedIn).user)
        assertEquals(user, store.user)
    }

    @Test
    fun `저장된 세션이 없으면 로그아웃 상태다`() = runTest {
        val (store, _, _) = store()

        store.restore()

        assertIs<AuthState.SignedOut>(store.state.value)
    }

    /**
     * 복원이 실패해도 오류 화면에 갇히면 안 된다 — 그러면 다시 로그인할 방법이 없다.
     */
    @Test
    fun `복원 실패는 로그인 화면으로 돌아간다`() = runTest {
        val gateway = FakeGateway()
        gateway.restoreError = AuthException.Network("끊김")
        val (store, _, _) = store(gateway)

        store.restore()

        assertIs<AuthState.SignedOut>(store.state.value)
    }

    @Test
    fun `구글 로그인이 성공하면 로그인 상태가 된다`() = runTest {
        val gateway = FakeGateway()
        gateway.signInResult = Result.success(session())
        val (store, _, _) = store(gateway)

        store.signInWithGoogle()

        assertEquals(user, (store.state.value as AuthState.SignedIn).user)
    }

    /** 사용자가 계정 선택 창을 닫은 것은 오류가 아니다. */
    @Test
    fun `취소는 오류가 아니라 로그아웃 복귀다`() = runTest {
        val (store, _, _) = store(provider = FakeIdentityProvider(identity = null))

        store.signInWithGoogle()

        assertIs<AuthState.SignedOut>(store.state.value)
    }

    @Test
    fun `서버가 거부하면 이유를 담은 실패 상태가 된다`() = runTest {
        val gateway = FakeGateway()
        gateway.signInResult = Result.failure(AuthException.Rejected("Unacceptable audience"))
        val (store, _, _) = store(gateway)

        store.signInWithGoogle()

        val failed = assertIs<AuthState.Failed>(store.state.value)
        assertTrue(failed.message.contains("Unacceptable audience"))
    }

    @Test
    fun `로그아웃하면 구글 세션까지 함께 끊는다`() = runTest {
        val gateway = FakeGateway(restored = session())
        val (store, _, provider) = store(gateway)
        store.restore()

        store.signOut()

        assertIs<AuthState.SignedOut>(store.state.value)
        assertEquals(1, gateway.signOutCount)
        assertEquals(1, provider.signOutCount)
    }

    /**
     * Supabase 로그아웃이 실패해도 로컬 로그아웃은 되어야 한다.
     * 여기서 멈추면 사용자가 로그아웃할 방법이 없어진다.
     */
    @Test
    fun `서버 로그아웃이 실패해도 로그아웃된다`() = runTest {
        val gateway = FakeGateway(restored = session())
        gateway.signOutError = AuthException.Network("끊김")
        val (store, _, provider) = store(gateway)
        store.restore()

        store.signOut()

        assertIs<AuthState.SignedOut>(store.state.value)
        assertEquals(1, provider.signOutCount)
    }
}

File: apps/mobile/packages/drop_core/test/auth_store_test.dart
import 'package:drop_core/drop_core.dart';
import 'package:test/test.dart';

/// DropCore `AuthStoreTests.swift` 포팅.
/// 실제 Google / Supabase 없이 게이트웨이를 갈아끼워 상태 전이를 검증한다.
void main() {
  (AuthStore, FakeAuthenticationGateway, FakeGoogleIdentityProvider) makeStore({
    FakeAuthenticationGateway? gateway,
    FakeGoogleIdentityProvider? identity,
  }) {
    final g = gateway ?? FakeAuthenticationGateway();
    final i = identity ?? FakeGoogleIdentityProvider();
    return (AuthStore(gateway: g, identityProvider: i), g, i);
  }

  group('인증 상태 전이', () {
    test('시작 상태는 판단 보류다', () {
      final (store, _, _) = makeStore();
      expect(store.state, const AuthUndetermined());
    });

    test('기존 세션이 있으면 복원한다', () async {
      final gateway = FakeAuthenticationGateway()
        ..currentUser =
            const DropUser(id: 'u1', email: 'bruce@intellieffect.com');
      final (store, _, _) = makeStore(gateway: gateway);

      await store.restore();

      expect(
        store.state,
        const AuthSignedIn(DropUser(id: 'u1', email: 'bruce@intellieffect.com')),
      );
    });

    test('세션이 없으면 로그아웃 상태다', () async {
      final (store, _, _) = makeStore();

      await store.restore();

      expect(store.state, const AuthSignedOut());
    });

    test('로그인에 성공하면 사용자를 노출한다', () async {
      final identity = FakeGoogleIdentityProvider()
        ..identity = const GoogleIdentity(idToken: 'id', accessToken: 'access');
      final gateway = FakeAuthenticationGateway()
        ..signInError = null
        ..signInUser = const DropUser(id: 'u2', email: 'a@b.c');
      final (store, _, _) = makeStore(gateway: gateway, identity: identity);

      await store.signInWithGoogle();

      expect(store.state, const AuthSignedIn(DropUser(id: 'u2', email: 'a@b.c')));
      expect(gateway.receivedIdToken, 'id');
      expect(gateway.receivedAccessToken, 'access');
    });

    /// 사용자가 로그인 창을 닫은 것은 오류가 아니다.
    test('사용자가 취소하면 오류가 아니라 로그아웃으로 되돌린다', () async {
      final identity = FakeGoogleIdentityProvider()..identity = null;
      final (store, _, _) = makeStore(identity: identity);

      await store.signInWithGoogle();

      expect(store.state, const AuthSignedOut());
    });

    test('로그인이 실패하면 메시지를 남긴다', () async {
      final identity = FakeGoogleIdentityProvider()
        ..identity = const GoogleIdentity(idToken: 'id', accessToken: null);
      final gateway = FakeAuthenticationGateway()..signInError = 'boom';
      final (store, _, _) = makeStore(gateway: gateway, identity: identity);

      await store.signInWithGoogle();

      expect(store.state, isA<AuthFailed>());
    });

    test('로그아웃하면 Google 세션까지 함께 끊는다', () async {
      final identity = FakeGoogleIdentityProvider()
        ..identity = const GoogleIdentity(idToken: 'id', accessToken: null);
      final gateway = FakeAuthenticationGateway()
        ..signInError = null
        ..signInUser = const DropUser(id: 'u3', email: null);
      final (store, _, _) = makeStore(gateway: gateway, identity: identity);
      await store.signInWithGoogle();

      await store.signOut();

      expect(store.state, const AuthSignedOut());
      expect(gateway.signOutCallCount, 1);
      expect(identity.signOutCallCount, 1);
    });

    /// 실패한 뒤 다시 시도하면 이전 오류 메시지가 남아 있으면 안 된다.
    test('재시도하면 이전 오류 표시가 사라진다', () async {
      final identity = FakeGoogleIdentityProvider()
        ..identity = const GoogleIdentity(idToken: 'id', accessToken: null);
      final gateway = FakeAuthenticationGateway()..signInError = 'boom';
      final (store, _, _) = makeStore(gateway: gateway, identity: identity);
      await store.signInWithGoogle();

      gateway
        ..signInError = null
        ..signInUser = const DropUser(id: 'u4', email: null);
      await store.signInWithGoogle();

      expect(store.state, const AuthSignedIn(DropUser(id: 'u4', email: null)));
    });
  });
}

class FakeAuthenticationGateway implements AuthenticationGateway {
  @override
  DropUser? currentUser;

  DropUser? signInUser;
  String? signInError = 'boom';
  String? receivedIdToken;
  String? receivedAccessToken;
  int signOutCallCount = 0;

  @override
  Future<DropUser> signIn({required String idToken, String? accessToken}) async {
    receivedIdToken = idToken;
    receivedAccessToken = accessToken;
    final error = signInError;
    if (error != null) throw StateError(error);
    return signInUser!;
  }

  @override
  Future<void> signOut() async {
    signOutCallCount += 1;
    currentUser = null;
  }
}

class FakeGoogleIdentityProvider implements GoogleIdentityProvider {
  GoogleIdentity? identity;
  int signOutCallCount = 0;

  @override
  Future<GoogleIdentity?> signIn() async => identity;

  @override
  void signOut() {
    signOutCallCount += 1;
  }
}
